### PR TITLE
Feature firebase

### PR DESCRIPTION
**Pull request by Julianna-Venter**
This PR includes:
- base UI **that is not complete**
- login and signup via firebase
- form validation and error handling
- adding the new user to firestore when registering a new user
- protected routes and routing when signed in or not
- getting a user's data when signed in and displaying it (just in a console log for now) in the home component, so that it can later be passed down to child components

Just a note: please double check my use of observables, I use a few between user authentication, and then to gather the user information in the home component, as a lot of these methods are async. Please call me out if something alludes to bad practices, I tried to avoid nested subscriptions but could've missed an alternative method or a better way to do this.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -36,6 +36,7 @@ yarn-error.log
 /libpeerconnection.log
 testem.log
 /typings
+**environments
```
Comment (yanevdw) [inline .gitignore:39]: Nice! 🔒

File: src/app/app.config.ts [hunk 2]
```diff
@@ -1,8 +1,27 @@
...
 
+const firebaseConfig = {
+  apiKey: FIRESTORE_API_KEY.apiKey,
+  authDomain: 'angular-book-tracker-686b9.firebaseapp.com',
+  projectId: 'angular-book-tracker-686b9',
+  storageBucket: 'angular-book-tracker-686b9.appspot.com',
+  messagingSenderId: '706862192768',
+  appId: '1:706862192768:web:cab46e521fb925ede86e65',
+};
+
```
Comment (yanevdw) [inline src/app/app.config.ts:17]: You could potentially hide your entire `firebaseConfig` in  `environments`. However, this is just a suggestion.

File: src/app/components/home/book-page/dnf/dnf.component.html
Comment (yanevdw) [inline src/app/components/home/book-page/dnf/dnf.component.html:1]: I am assuming this component still needs to be implemented?

**Review by yanevdw — APPROVED** [review]
The firebase config, setup and auth looks good. 

I assume all the components that still have the default "component works!" in the HTML files still have to be implemented?

File: src/assets/database.png
Comment (tanadendere) [inline src/assets/database.png:1]: I see that both in `ProfileStats` and in `Book` you note the `Likes` and `Dislikes` attributes as array of genres/ subjects. But I do not see a genre/ subjects collection in your database design? 

And also I am not sure how they are related to the Book collection? Is it maybe from that book's genres which ones do they enjoy? If it is, then I would think that ones in ProfileStates would be derived information.

File: src/app/components/home/book-page/dnf/dnf.component.html
Comment (Julianna-Venter) [reply src/app/components/home/book-page/dnf/dnf.component.html:1]: yeah

File: src/app/components/home/profile-stats/log-out/log-out.component.ts
Comment (tanadendere) [inline src/app/components/home/profile-stats/log-out/log-out.component.ts:1]: I really like that you separated this into its own component 🤌🏾

File: src/app/auth/auth.guard.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+import { take } from 'rxjs/operators';
+
+export const authGuard: CanActivateFn = (route, state) => {
+  const authService = inject(AuthService);
+  const router = inject(Router);
+
+  authService.user$.pipe(take(1)).subscribe((user) => {
+    if (user) {
+      authService.currentUserSig.set({
+        //! is added because it is pretty much guaranteed that the user is logged in at this point
```
Comment (tanadendere) [inline src/app/auth/auth.guard.ts:14]: I would suggest using a ternary like` user.email ? user.email : ""`. Incase the user is logged in but the api returns null values for the user.

File: src/app/services/users-firebase.service.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        read: [],
+        tbr: [],
+      },
+      profilestats: {
+        stats: [],
+      },
+    });
+  }
+
+  getUser(userEmail: string): Observable<FirestoreUser[]> {
```
Comment (tanadendere) [inline src/app/services/users-firebase.service.ts:36]: Clean 🤌🏾🔥

**Review by tanadendere — CHANGES_REQUESTED** [review]
Well done Jules, excellent work as always! It is coming along really well. I just have the couple changes from the comments.

File: src/assets/database.png
Comment (Julianna-Venter) [reply src/assets/database.png:1]: these genres/subjects are ones that I get from the API, which is then added just into an array in their respective subarrays. The stats one will be derived of all the user's liked ones and the ones that appear most often

File: src/app/components/log-in/log-in.component.ts [hunk 5]
```diff
@@ -1,22 +1,67 @@
...
   onSubmit() {
-    // TODO: Use EventEmitter with form value
-    console.warn(this.loginForm.value);
+    // TODO: when this is implememted you need to have a spinner to indicate that the form is being submitted
+    console.log(this.loginForm.value);
+    const rawForm = this.loginForm.getRawValue();
+
+    this.authService
+      .login(rawForm.email ?? '', rawForm.password ?? '')
+      .subscribe({
```
Comment (Musa-Mabasa) [inline src/app/components/log-in/log-in.component.ts:53]: Remember to destroy subscriptions.

File: src/app/components/sign-up/sign-up.component.ts [hunk 6]
```diff
@@ -1,12 +1,127 @@
...
+      console.warn(this.signupForm.value);
+      const rawForm = this.signupForm.getRawValue();
+
+      this.authService
+        .register(
+          rawForm.email ?? '',
+          rawForm.username ?? '',
+          rawForm.password ?? ''
+        )
+        .subscribe({
```
Comment (Musa-Mabasa) [inline src/app/components/sign-up/sign-up.component.ts:97]: Same here with destroying the subscription

File: src/app/components/log-in/log-in.component.ts (hunk 5 above)
Comment (Julianna-Venter) [reply src/app/components/log-in/log-in.component.ts:53]: yeah, spoke with Tim, this will change when changing to NgRx, which I am currently implementing

**Review by tanadendere — APPROVED** [review]
Nice!

**Pull request merged**